docs/proposals/20240807-in-place-updates.md

```diff
@@ -0,0 +1,484 @@
+5. Machine Controller: make machine is updated.
+
+The following sections dive deep into these steps, zooming in into the different component interactions and defining how the main error cases are handled.
+
+### Defining the Update Plan
+
+```mermaid
+sequenceDiagram
+    participant Operator
+
+    box Management Cluster
+        participant apiserver as kube-api server
+        participant capi as CP/MD Controller
+        participant hook as External updater 1
+        participant hook2 as External updater 2
+    end
+
+    Operator->>+apiserver: make changes to CP/MD <br>(ExternalUpdate strategy)
+    apiserver->>+capi: Notify changes
+    apiserver->>-Operator: OK
+    capi->>+hook: Can update?
+    hook->>capi: Set of changes
+    capi->>+hook2: Can update?
+    hook2->>capi: Set of changes
+    alt all changes covered?
+        capi->>apiserver: Define Update Plan
+    else
+        alt fallback strategy?
+            capi->>apiserver: Re-create machines
+        else
+            capi->>apiserver: Marked update as failed
+        end
+    end
+```
+
+Both `KCP` and `MachineDeployment` controllers follow a similar pattern around updates, they first detect if an update is required and then based on the configured strategy follow the appropiate update logic (note that today there is only one valid strategy, `RollingUpdate`).
+
+With `ExternalUpdate` strategy, CAPI controllers will compute the set of desired changes and iterate over the registered external updaters, requesting through the Runtime Hook the set of changes each updater can handle. The changes supported by an updater can be the complete set of desired changes, a subset of them or an empty set, signaling it cannot handle any of the desired changes.
+
+On each iteration, the controller will substract from the set of desired changes the subset returned by the updater. 
+
+If this set is reduced to zero, then CAPI will determine that the update can be performed using the external strategy. CAPI will define the update plan as a list of sequential external updaters in a particular order and proceed to execute it. The update plan will be stored in the Machine object as an array of strings (the names of the selected external updaters).
+
+If after iterating over all external updaters the remaining set still contains uncovered changes, CAPI will determine the desired state cannot be reached through external updaters. If a fallback rolling update strategy has been configured (this is optional), CAPI will replace the machines. If no fallback strategy is configured, we will surface the issue in the resource status. Machines will remain unchanged and the desired state won't be reached unless remediated by the user. Depending on the scenario, users can: ammend the desired state to something that the registered updaters can cover, register additional updaters capable of handling the desired changes or simply enable the fallback strategy.
```

Comment:
> If after iterating over all external updaters the remaining set still contains uncovered changes

How do we envision this to take place? Diffing each field?

Reply:
I envision something like capi would generate a set with all the fields that are changing for an object by diffing current state with desired state. Then as it iterates over the updaters, it would remove fields from the set. If it finish iterating over the updaters and there are still fields left in the set, then the update can't be performed in-place.